File: scratchpad/gleam/os/src/gloss_ffi.mjs
import {
  new$ as newDict,
  insert as insertDict,
} from "../gleam_stdlib/gleam/dict.mjs";
import { Result$Error, Result$Ok } from "./gleam.mjs";

let os = globalThis?.process?.platform
  || globalThis?.navigator?.userAgentData?.platform?.toLowerCase()
  || "unknown";

if (os === "windows") {
  os = "win32"
} else if (os === "iOS" || os === "macOS") {
  os = "darwin";
} else if (os === "android" || os === "chrome os") {
  os = "linux"
}

export function is_windows() {
  return os == "win32";
}

export function system_name() {
  return os;
}

export function environment_get(key) {
  const value = globalThis?.process?.env[key];

  if (value === undefined) {
    return Result$Error(undefined);
  } else {
    return Result$Ok(value);
  }
}

export function environment_set(key, value) {
  if (globalThis.process?.env) {
    globalThis.process.env[key] = value;
  }
}

export function environment_unset(key) {
  if (globalThis.process?.env) {
    delete process.env[key];
  }
}

export function environment_all() {
  const environmentVariables = globalThis.process?.env ?? {};
  let result = newDict();
  for (const key in environmentVariables) {
    if (Object.hasOwn(environmentVariables, key)) {
      result = insertDict(result, key, environmentVariables[key]);
    }
  }

  return result;
}

File: scratchpad/gleam/os/src/gloss_ffi.erl
-module(gloss_ffi).

-export([
    is_windows/0, environment_get/1, environment_set/2, environment_unset/1,
    environment_all/0, system_name/0
]).

system_name() ->
    case os:type() of
        {win32, _} -> ~"win32";
        {_, Name} -> erlang:atom_to_binary(Name)
    end.

is_windows() ->
    case os:type() of
        {win32, _} -> true;
        _ -> false
    end.

environment_get(Key) ->
    case os:getenv(unicode:characters_to_list(Key)) of
        false -> {error, nil};
        Value -> {ok, unicode:characters_to_binary(Value)}
    end.

environment_set(Key, Value) ->
    os:putenv(unicode:characters_to_list(Key), unicode:characters_to_list(Value)),
    nil.

environment_unset(Key) ->
    os:unsetenv(unicode:characters_to_list(Key)),
    nil.

environment_all() ->
    BinVars = lists:map(fun(VarString) ->
        [VarName, VarVal] = string:split(VarString, "="),
        {unicode:characters_to_binary(VarName), unicode:characters_to_binary(VarVal)}
    end, os:getenv()),
    maps:from_list(BinVars).
